File: spreadsheet_llm/cell_utils.py
"""Cell type detection, format parsing, and semantic type classification.

Implements the 9 semantic types from SpreadsheetLLM paper Section 3.4:
Year, Integer, Float, Percentage, Scientific, Date, Time, Currency, Email.
"""

from __future__ import annotations

import re
from typing import Tuple

from openpyxl.cell.cell import Cell
from openpyxl.utils import get_column_letter

# Pre-compiled regex patterns (module-level for performance)
EMAIL_REGEX = re.compile(r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$")
CURRENCY_REGEX = re.compile(r"[$€£¥]")
SCIENTIFIC_REGEX = re.compile(r"E[+-]", re.IGNORECASE)
DATE_KEYWORDS = re.compile(r"(yyyy|yy|mmmm|mmm|dd|ddd|dddd)", re.IGNORECASE)
TIME_KEYWORDS = re.compile(r"(hh?|ss?|am/pm|a/p)", re.IGNORECASE)
YEAR_ONLY_REGEX = re.compile(r"^y{2,4}$", re.IGNORECASE)
NUMERIC_NFS_REGEX = re.compile(r"^[#0,]+(\.[#0]+)?$")

# Style fingerprint type
StyleFingerprint = Tuple[int, int, int, int]


def infer_cell_data_type(cell: Cell) -> str:
    """Infer the data type of a cell from openpyxl metadata and value.

    Returns one of: empty, text, numeric, boolean, datetime, email, error.
    """
    if cell.value is None:
        return "empty"

    # Check email before falling through to text
    if isinstance(cell.value, str) and EMAIL_REGEX.match(cell.value):
        return "email"

    dt = cell.data_type
    if dt == "s":
        return "text"
    if dt == "n":
        return "numeric"
    if dt == "b":
        return "boolean"
    if dt == "d":
        return "datetime"
    if dt == "e":
        return "error"
    if dt == "f":
        # Formula — infer from cached value
        v = cell.value
        if isinstance(v, str):
            return "text"
        if isinstance(v, (int, float)):
            return "numeric"
        if isinstance(v, bool):
            return "boolean"
        return "formula"
    return "text"  # fallback


def get_number_format_string(cell: Cell) -> str:
    """Return the raw Number Format String for a cell."""
    try:
        nfs = cell.number_format
        if nfs is None or nfs == "":
            return "General"
        return str(nfs)
    except Exception:
        return "General"


def categorize_number_format(nfs: str, cell: Cell) -> str:
    """Categorize a number format string into a broad category.

    Returns one of: general, currency, percentage, scientific, fraction,
    date_custom, time_custom, datetime_custom, integer, float,
    other_numeric, not_applicable.
    """
    cell_type = infer_cell_data_type(cell)
    if cell_type not in ("numeric", "datetime"):
        return "not_applicable"

    if nfs is None or nfs.lower() == "general":
        return "datetime_general" if cell_type == "datetime" else "general"

    if nfs == "@" or nfs.lower() == "text":
        return "text_format"

    if CURRENCY_REGEX.search(nfs):
        return "currency"
    if "%" in nfs:
        return "percentage"
    if SCIENTIFIC_REGEX.search(nfs):
        return "scientific"
    if "#" in nfs and "/" in nfs and "?" in nfs:
        return "fraction"

    nfs_lower = nfs.lower()
    is_date = bool(DATE_KEYWORDS.search(nfs_lower))
    is_time = bool(TIME_KEYWORDS.search(nfs_lower))

    # Bare "m" could be month or minute — disambiguate
    if "m" in nfs_lower and not is_date and not is_time:
        if any(k in nfs_lower for k in ("h", "s")):
            is_time = True
        elif re.fullmatch(r"m{1,5}", nfs_lower):
            is_date = True

    if is_date and is_time:
        return "datetime_custom"
    if is_date:
        return "date_custom"
    if is_time:
        return "time_custom"

    if cell_type == "numeric":
        if nfs in ("0", "#,##0"):
            return "integer"
        if nfs in ("0.00", "#,##0.00", "0.0", "#,##0.0"):
            return "float"
        if NUMERIC_NFS_REGEX.match(nfs):
            return "other_numeric"

    if cell_type == "datetime":
        return "other_date"

    return "not_applicable"


def detect_semantic_type(cell: Cell) -> str:
    """Detect the semantic type of a cell per the paper's 9 categories.

    Returns one of: year, integer, float, percentage, scientific,
    date, time, currency, email, text, empty, boolean, error.
    """
    data_type = infer_cell_data_type(cell)

    if data_type == "empty":
        return "empty"
    if data_type == "email":
        return "email"
    if data_type == "boolean":
        return "boolean"
    if data_type == "error":
        return "error"

    nfs = get_number_format_string(cell)
    category = categorize_number_format(nfs, cell)
    nfs_lower = nfs.lower()

    if category == "percentage":
        return "percentage"
    if category == "currency":
        return "currency"
    if category == "scientific":
        return "scientific"
    if category in ("date_custom", "datetime_custom", "datetime_general", "other_date"):
        # Check if it's year-only format
        if YEAR_ONLY_REGEX.match(nfs_lower):
            return "year"
        return "date"
    if category == "time_custom":
        return "time"

    if data_type == "numeric":
        if isinstance(cell.value, int) or category == "integer":
            return "integer"
        if isinstance(cell.value, float) or category in ("float", "other_numeric"):
            return "float"
        return "integer"  # numeric fallback

    return "text"


def get_cell_style_fingerprint(cell: Cell) -> StyleFingerprint:
    """Return a hashable style fingerprint for boundary comparison.

    Uses attribute hashing (works in both in-memory and loaded workbooks).
    """
    try:
        font = cell.font
        f_key = (font.bold, font.italic, font.underline, font.sz,
                 str(font.color.rgb) if font.color and hasattr(font.color, 'rgb') and font.color.rgb else None)
    except Exception:
        f_key = None

    try:
        border = cell.border
        b_key = (border.left.style, border.right.style,
                 border.top.style, border.bottom.style)
    except Exception:
        b_key = None

    try:
        fill = cell.fill
        fill_key = (fill.patternType,
                    str(fill.fgColor.rgb) if fill.fgColor and hasattr(fill.fgColor, 'rgb') and fill.fgColor.rgb else None)
    except Exception:
        fill_key = None

    try:
        align = cell.alignment
        a_key = (align.horizontal, align.vertical)
    except Exception:
        a_key = None

    return (hash(f_key), hash(b_key), hash(fill_key), hash(a_key))


def cell_coord(row: int, col: int) -> str:
    """Convert (row, col) to Excel coordinate like 'A1'."""
    return f"{get_column_letter(col)}{row}"


def split_cell_ref(ref: str) -> tuple[str, int]:
    """Split 'AB12' into ('AB', 12)."""
    col_str = ""
    row_str = ""
    for ch in ref:
        if ch.isalpha():
            col_str += ch
        else:
            row_str += ch
    return col_str, int(row_str)

File: SpreadsheetLLM.Core/CellUtils.cs
using System;
using System.Text.RegularExpressions;
using SpreadsheetLLM.Core.Models;

namespace SpreadsheetLLM.Core
{
    /// <summary>
    /// Cell type detection, format parsing, and semantic type classification.
    /// Port of spreadsheet_llm/cell_utils.py.
    ///
    /// Implements the 9 semantic types from SpreadsheetLLM paper Section 3.4:
    /// Year, Integer, Float, Percentage, Scientific, Date, Time, Currency, Email.
    /// </summary>
    public static class CellUtils
    {
        // Pre-compiled regex patterns (static for performance — module-level equivalent)
        private static readonly Regex EmailRegex =
            new Regex(@"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$",
                RegexOptions.Compiled | RegexOptions.IgnoreCase);

        private static readonly Regex CurrencyRegex =
            new Regex(@"[$€£¥]", RegexOptions.Compiled);

        private static readonly Regex ScientificRegex =
            new Regex(@"E[+-]", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        private static readonly Regex DateKeywords =
            new Regex(@"(yyyy|yy|mmmm|mmm|dd|ddd|dddd)", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        private static readonly Regex TimeKeywords =
            new Regex(@"(hh?|ss?|am/pm|a/p)", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        private static readonly Regex YearOnlyRegex =
            new Regex(@"^y{2,4}$", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        private static readonly Regex NumericNfsRegex =
            new Regex(@"^[#0,]+(\.[#0]+)?$", RegexOptions.Compiled);

        /// <summary>
        /// Infer the basic data type of a cell.
        /// Returns: "empty" | "text" | "numeric" | "boolean" | "datetime" | "email" | "error" | "formula"
        /// </summary>
        public static string InferCellDataType(CellData cell)
        {
            if (cell.Value == null)
                return "empty";

            // Check email before falling through to text
            if (!cell.IsFormula && EmailRegex.IsMatch(cell.Value))
                return "email";

            if (cell.IsFormula)
            {
                // For formula cells, infer from computed value
                if (cell.ComputedValue == null)
                    return "formula";
                if (double.TryParse(cell.ComputedValue, out _))
                    return "numeric";
                if (bool.TryParse(cell.ComputedValue, out _))
                    return "boolean";
                return "text";
            }

            // Try to infer from value string
            if (cell.Value == "TRUE" || cell.Value == "FALSE")
                return "boolean";
            if (double.TryParse(cell.Value, System.Globalization.NumberStyles.Any,
                System.Globalization.CultureInfo.InvariantCulture, out _))
                return "numeric";

            // Check datetime via number format
            var nfs = cell.NumberFormat.ToLowerInvariant();
            if (DateKeywords.IsMatch(nfs) || TimeKeywords.IsMatch(nfs))
                return "datetime";

            return "text";
        }

        /// <summary>Returns the raw Number Format String for a cell.</summary>
        public static string GetNumberFormatString(CellData cell)
        {
            var nfs = cell.NumberFormat;
            if (string.IsNullOrEmpty(nfs))
                return "General";
            return nfs;
        }

        /// <summary>
        /// Categorize a number format string into a broad category.
        /// Returns: general | currency | percentage | scientific | fraction |
        ///          date_custom | time_custom | datetime_custom | integer | float |
        ///          other_numeric | not_applicable | text_format
        /// </summary>
        public static string CategorizeNumberFormat(string nfs, CellData cell)
        {
            var cellType = InferCellDataType(cell);
            if (cellType != "numeric" && cellType != "datetime")
                return "not_applicable";

            if (string.IsNullOrEmpty(nfs) || nfs.Equals("General", StringComparison.OrdinalIgnoreCase))
                return cellType == "datetime" ? "datetime_general" : "general";

            if (nfs == "@" || nfs.Equals("text", StringComparison.OrdinalIgnoreCase))
                return "text_format";

            if (CurrencyRegex.IsMatch(nfs))
                return "currency";
            if (nfs.Contains("%"))
                return "percentage";
            if (ScientificRegex.IsMatch(nfs))
                return "scientific";
            if (nfs.Contains("#") && nfs.Contains("/") && nfs.Contains("?"))
                return "fraction";

            var nfsLower = nfs.ToLowerInvariant();
            bool isDate = DateKeywords.IsMatch(nfsLower);
            bool isTime = TimeKeywords.IsMatch(nfsLower);

            // Bare "m" disambiguation: month vs minute
            if (nfsLower.Contains("m") && !isDate && !isTime)
            {
                if (nfsLower.Contains("h") || nfsLower.Contains("s"))
                    isTime = true;
                else if (Regex.IsMatch(nfsLower, @"^m{1,5}$"))
                    isDate = true;
            }

            if (isDate && isTime) return "datetime_custom";
            if (isDate) return "date_custom";
            if (isTime) return "time_custom";

            if (cellType == "numeric")
            {
                if (nfs == "0" || nfs == "#,##0") return "integer";
                if (nfs == "0.00" || nfs == "#,##0.00" || nfs == "0.0" || nfs == "#,##0.0") return "float";
                if (NumericNfsRegex.IsMatch(nfs)) return "other_numeric";
            }

            if (cellType == "datetime")
                return "other_date";

            return "not_applicable";
        }

        /// <summary>
        /// Detect the semantic type of a cell per the paper's 9 categories.
        /// Returns: year | integer | float | percentage | scientific | date | time | currency | email | text | empty | boolean | error
        /// </summary>
        public static string DetectSemanticType(CellData cell)
        {
            var dataType = InferCellDataType(cell);

            if (dataType == "empty") return "empty";
            if (dataType == "email") return "email";
            if (dataType == "boolean") return "boolean";
            if (dataType == "error") return "error";

            var nfs = GetNumberFormatString(cell);
            var category = CategorizeNumberFormat(nfs, cell);
            var nfsLower = nfs.ToLowerInvariant();

            if (category == "percentage") return "percentage";
            if (category == "currency") return "currency";
            if (category == "scientific") return "scientific";

            if (category == "date_custom" || category == "datetime_custom" ||
                category == "datetime_general" || category == "other_date")
            {
                if (YearOnlyRegex.IsMatch(nfsLower))
                    return "year";
                return "date";
            }

            if (category == "time_custom") return "time";

            if (dataType == "numeric")
            {
                // Check if the value string represents an integer
                if (cell.Value != null && !cell.IsFormula)
                {
                    if (long.TryParse(cell.Value, out _))
                        return "integer";
                    if (category == "integer")
                        return "integer";
                    return "float";
                }
                if (category == "integer") return "integer";
                if (category == "float" || category == "other_numeric") return "float";
                return "integer"; // numeric fallback
            }

            return "text";
        }

        /// <summary>
        /// Returns a stable integer fingerprint for a cell's style.
        /// Used for boundary detection — two cells with the same style get the same fingerprint.
        /// Uses HashCode.Combine on actual attribute values (not object identity).
        /// </summary>
        public static int GetStyleFingerprint(CellData cell)
        {
            // Combine all style-relevant fields into a single hash
            int fontKey = CombineHash(
                cell.FontBold.GetHashCode(),
                cell.FontItalic.GetHashCode(),
                (cell.FontColor ?? "").GetHashCode()
            );
            int borderKey = (cell.BorderBottomStyle ?? "").GetHashCode();
            int fillKey = CombineHash(
                (cell.FillPatternType ?? "").GetHashCode(),
                (cell.FillFgColor ?? "").GetHashCode()
            );
            int alignKey = (cell.AlignmentHorizontal ?? "").GetHashCode();

            return CombineHash(fontKey, borderKey, fillKey, alignKey);
        }

        private static int CombineHash(int h1, int h2) =>
            unchecked(((h1 << 5) + h1) ^ h2);

        private static int CombineHash(int h1, int h2, int h3) =>
            CombineHash(CombineHash(h1, h2), h3);

        private static int CombineHash(int h1, int h2, int h3, int h4) =>
            CombineHash(CombineHash(h1, h2), CombineHash(h3, h4));

        /// <summary>Converts (row, col) 1-based integers to an Excel cell reference like "A1".</summary>
        public static string CellCoord(int row, int col)
        {
            return ColumnLetter(col) + row.ToString();
        }

        /// <summary>Converts a 1-based column number to a column letter string (e.g., 1 → "A", 27 → "AA").</summary>
        public static string ColumnLetter(int col)
        {
            var result = "";
            while (col > 0)
            {
                col--;
                result = (char)('A' + col % 26) + result;
                col /= 26;
            }
            return result;
        }

        /// <summary>Converts a column letter string to a 1-based column number (e.g., "A" → 1, "AA" → 27).</summary>
        public static int ColumnNumber(string colLetter)
        {
            int result = 0;
            foreach (var ch in colLetter.ToUpperInvariant())
            {
                result = result * 26 + (ch - 'A' + 1);
            }
            return result;
        }

        /// <summary>Splits a cell reference like "AB12" into ("AB", 12).</summary>
        public static (string colLetter, int row) SplitCellRef(string cellRef)
        {
            int i = 0;
            while (i < cellRef.Length && char.IsLetter(cellRef[i]))
                i++;
            var col = cellRef.Substring(0, i).ToUpperInvariant();
            var row = int.Parse(cellRef.Substring(i));
            return (col, row);
        }
    }
}
